openfisca_core/entities.py

```diff
@@ -38,20 +143,20 @@ def to_json(cls):
     # Calculations
 
     def check_variable_defined_for_entity(self, variable_name):
-        if not (self.simulation.get_variable_entity(variable_name) == self):
-            variable_entity = self.simulation.get_variable_entity(variable_name)
-            raise Exception(
+        variable_entity = self.simulation.tax_benefit_system.get_column(variable_name, check_existence = True).entity
+        if not isinstance(self, variable_entity):
+            raise ValueError(
                 "Variable {} is not defined for {} but for {}".format(
-                    variable_name, self.key, variable_entity.key)
+                    variable_name, self.plural, variable_entity.plural)
                 )
 
     def check_array_compatible_with_entity(self, array):
         if not self.count == array.size:
-            raise Exception("Input {} is not a valid value for the entity {}".format(array, self.key))
+            raise ValueError("Input {} is not a valid value for the entity {}".format(array, self.key))
 
     def check_role_validity(self, role):
         if role is not None and not type(role) == Role:
-            raise Exception("{} is not a valid role".format(role))
+            raise ValueError("{} is not a valid role".format(role))
```

Comment:
`check_role_validity` is not called in the entity initialization process.
This was just an opportunistic fix.